# Data Preparation

[Original Notebook source from *Data Science: Introduction to Machine Learning for Data Science Python and Machine Learning Studio by Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## Exploring `DataFrame` information

> **Learning goal:** By the end of this subsection, you should be comfortable finding general information about the data stored in pandas DataFrames.

Once you have loaded your data into pandas, it will most likely be in a `DataFrame`. But if the dataset in your `DataFrame` contains 60,000 rows and 400 columns, how do you even start to understand what you're working with? Fortunately, pandas offers some handy tools to quickly get an overview of a `DataFrame`, as well as to examine the first few and last few rows.

To explore this functionality, we will import the Python scikit-learn library and use a classic dataset that every data scientist has encountered countless times: British biologist Ronald Fisher's *Iris* dataset, featured in his 1936 paper "The use of multiple measurements in taxonomic problems":


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
We have loaded the Iris Dataset into the variable `iris_df`. Before analyzing the data, it is helpful to know the number of data points and the overall size of the dataset. Understanding the volume of data we are working with is important.


In [2]:
iris_df.shape

(150, 4)

So, we are working with 150 rows and 4 columns of data. Each row corresponds to one data point, and each column represents a single feature associated with the data frame. Essentially, there are 150 data points, each containing 4 features.

`shape` is an attribute of the dataframe, not a function, which is why it doesn't have parentheses at the end.


### `DataFrame.columns`
Now let's delve into the 4 columns of data. What does each of them specifically represent? The `columns` attribute will provide us with the names of the columns in the dataframe.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

As we can see, there are four(4) columns. The `columns` attribute tells us the name of the columns and basically nothing else. This attribute assumes importance when we want to identify the features a dataset contains.


### `DataFrame.info`
The size of the dataset (provided by the `shape` attribute) and the names of the features or columns (provided by the `columns` attribute) give us some initial insights into the dataset. Now, we might want to explore the dataset in more detail. The `DataFrame.info()` function is particularly helpful for this.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


From here, we can make a few observations:
1. The DataType of each column: In this dataset, all the data is stored as 64-bit floating-point numbers.
2. Number of Non-Null values: Handling null values is a crucial step in data preparation. This will be addressed later in the notebook.


### DataFrame.describe()
Suppose we have a large amount of numerical data in our dataset. Univariate statistical calculations like mean, median, quartiles, etc., can be performed on each column individually. The `DataFrame.describe()` function gives us a statistical summary of the numerical columns in a dataset.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

The output above shows the total number of data points, mean, standard deviation, minimum, lower quartile (25%), median (50%), upper quartile (75%), and the maximum value of each column.


### `DataFrame.head`
With all the functions and attributes mentioned above, we now have a high-level overview of the dataset. We know the total number of data points, the number of features, the data type of each feature, and the count of non-null values for each feature.

Now it's time to examine the data itself. Let's take a look at the first few rows (the initial data points) of our `DataFrame`:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

As the output here, we can see five (5) entries of the dataset. If we look at the index on the left, we find out that these are the first five rows.


### Exercise:

From the example above, it's clear that by default, `DataFrame.head` returns the first five rows of a `DataFrame`. In the code cell below, can you find a way to display more than five rows?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
Another way to view the data is from the end (instead of the beginning). The counterpart to `DataFrame.head` is `DataFrame.tail`, which returns the last five rows of a `DataFrame`:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

In practice, it is helpful to quickly inspect the first few rows or the last few rows of a `DataFrame`, especially when searching for outliers in ordered datasets.

All the functions and attributes demonstrated above with code examples assist in providing an overview of the data.

> **Takeaway:** Simply examining the metadata of a DataFrame or the first and last few rows can give you an instant understanding of the size, structure, and content of the data you're working with.


### Missing Data
Let's explore the concept of missing data. Missing data occurs when no value is stored in certain columns.

For example, imagine someone is self-conscious about their weight and chooses not to fill out the weight field in a survey. In this case, the weight value for that individual will be missing.

In real-world datasets, missing values are quite common.

**How Pandas Handles Missing Data**

Pandas manages missing values in two ways. The first method, which you've encountered in earlier sections, is `NaN`, or Not a Number. This is a special value defined by the IEEE floating-point specification, and it is specifically used to represent missing floating-point values.

For missing values in data types other than floats, pandas uses Python's `None` object. While it might seem confusing to encounter two different types of values that essentially indicate the same thing, there are solid programmatic reasons behind this design choice. In practice, this approach allows pandas to strike a good balance for most use cases. However, it's important to note that both `None` and `NaN` come with certain limitations regarding how they can be utilized.


### `None`: non-float missing data
Since `None` originates from Python, it cannot be used in NumPy and pandas arrays unless their data type is `'object'`. Keep in mind that NumPy arrays (and pandas data structures) are designed to hold only one type of data. This characteristic is what makes them incredibly powerful for handling large-scale data and performing computational tasks, but it also restricts their flexibility. These arrays must convert to the "lowest common denominator," which is the data type capable of accommodating all elements in the array. When `None` is present in the array, it indicates that you are working with Python objects.

To illustrate this, take a look at the following example array (pay attention to its `dtype`):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

The reality of upcasting data types comes with two side effects. First, operations will be executed at the level of interpreted Python code rather than compiled NumPy code. In essence, this means that any operations involving `Series` or `DataFrames` containing `None` will be slower. While this performance impact might not be noticeable in most cases, it could become problematic for large datasets.

The second side effect is a consequence of the first. Since `None` essentially pulls `Series` or `DataFrames` back into the realm of standard Python, using NumPy/pandas aggregation functions like `sum()` or `min()` on arrays containing a `None` value will typically result in an error:


In [10]:
example1.sum()

TypeError: ignored

**Key takeaway**: Addition (and other operations) between integers and `None` values is undefined, which can limit what you can do with datasets that contain them.


### `NaN`: missing float values

Unlike `None`, NumPy (and consequently pandas) supports `NaN` for efficient, vectorized operations and ufuncs. The downside is that any arithmetic operation involving `NaN` will always yield `NaN`. For example:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

The good news: aggregations run on arrays with `NaN` in them don't pop errors. The bad news: the results are not uniformly useful:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### Exercise:


In [11]:
# What happens if you add np.nan and None together?


Remember: `NaN` is just for missing floating-point values; there is no `NaN` equivalent for integers, strings, or Booleans.


### `NaN` and `None`: null values in pandas

Even though `NaN` and `None` can behave slightly differently, pandas is designed to handle them interchangeably. To illustrate this, let's look at a `Series` of integers:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### Exercise:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


In the process of converting data types to ensure uniformity in `Series` and `DataFrame`s, pandas allows missing values to interchangeably appear as `None` or `NaN`. Due to this design, it can be useful to think of `None` and `NaN` as two variations of "null" in pandas. In fact, some of the key methods for handling missing values in pandas reflect this concept in their names:

- `isnull()`: Creates a Boolean mask to identify missing values
- `notnull()`: The inverse of `isnull()`
- `dropna()`: Produces a filtered version of the data
- `fillna()`: Returns a copy of the data with missing values replaced or filled

These methods are essential to understand and become proficient with, so let's explore each one in detail.


### Detecting null values

Now that we understand the importance of missing values, we need to identify them in our dataset before addressing them.  
Both `isnull()` and `notnull()` are your main methods for detecting null data. Both return Boolean masks for your data.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Look closely at the output. Does any of it surprise you? While `0` is an arithmetic null, it's still a perfectly valid integer, and pandas treats it as such. `''` is a bit more nuanced. Although we used it in Section 1 to represent an empty string value, it is still a string object and not considered a null representation by pandas.

Now, let's approach this differently and use these methods in a way that's closer to how you'll apply them in practice. Boolean masks can be used directly as a ``Series`` or ``DataFrame`` index, which is helpful when working with specific missing (or present) values.

If we want the total count of missing values, we can simply sum the mask generated by the `isnull()` method.


In [19]:
example3.isnull().sum()

2

### Exercise:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**Key takeaway**: Both the `isnull()` and `notnull()` methods produce similar results when you use them in DataFrames: they show the results and the index of those results, which will help you enormously as you wrestle with your data.


### Handling Missing Data

> **Learning goal:** By the end of this subsection, you should understand how and when to replace or remove null values in DataFrames.

Machine Learning models cannot process missing data directly. Therefore, before feeding the data into the model, we need to address these missing values.

The way missing data is handled involves subtle trade-offs and can influence your final analysis as well as real-world outcomes.

There are two main approaches to dealing with missing data:

1.   Remove the row containing the missing value
2.   Replace the missing value with another value

We will explore both methods in detail, along with their advantages and disadvantages.


### Dropping null values

The amount of data we provide to our model directly impacts its performance. Dropping null values means reducing the number of data points, which in turn decreases the size of the dataset. Therefore, it is recommended to drop rows with null values when the dataset is sufficiently large.

Another scenario could be when a specific row or column contains a significant number of missing values. In such cases, they can be dropped because they wouldn't contribute much to our analysis, as most of the data for that row/column is missing.

Beyond detecting missing values, pandas offers a convenient way to remove null values from `Series` and `DataFrame`s. To see this in practice, let's revisit `example3`. The `DataFrame.dropna()` function is useful for dropping rows with null values.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Note that this should resemble your output from `example3[example3.notnull()]`. The difference here is that, instead of simply indexing the masked values, `dropna` has eliminated those missing values from the `Series` `example3`.

Since DataFrames are two-dimensional, they offer more flexibility for removing data.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(Did you notice that pandas converted two of the columns to floats to handle the `NaN`s?)

You can't remove a single value from a `DataFrame`; instead, you have to remove entire rows or columns. Depending on your specific needs, you might choose one approach over the other, and pandas provides options for both. In data science, columns typically represent variables while rows represent observations, so it's more common to drop rows of data. The default behavior of `dropna()` is to remove all rows that contain any null values:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

If necessary, you can drop NA values from columns. Use `axis=1` to do so:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

Notice that this approach can remove a significant amount of data that you might want to retain, especially in smaller datasets. What if you only want to drop rows or columns that contain several null values or even just entirely null values? You can configure this behavior in `dropna` using the `how` and `thresh` parameters.

By default, `how='any'` (if you'd like to verify this or explore other parameters of the method, run `example4.dropna?` in a code cell). Alternatively, you can set `how='all'` to drop only rows or columns that are completely filled with null values. Let's expand our example `DataFrame` to observe this in action in the next exercise.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> Key takeaways: 
1. Dropping null values is advisable only if the dataset is sufficiently large.
2. Entire rows or columns can be removed if most of their data is missing.
3. The `DataFrame.dropna(axis=)` method is useful for eliminating null values. The `axis` argument determines whether rows or columns are dropped.
4. The `how` argument can also be utilized. By default, it is set to `any`, meaning it will drop rows/columns containing any null values. It can be set to `all` to specify that only rows/columns where all values are null will be dropped.


### Exercise:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


The `thresh` parameter gives you finer-grained control: you set the number of *non-null* values that a row or column needs to have in order to be kept:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

Here, the first and last row have been dropped, because they contain only two non-null values.


### Filling null values

Sometimes, it makes sense to replace missing values with ones that could be valid. There are several techniques to handle null values. The first involves using Domain Knowledge (expertise in the subject matter related to the dataset) to approximate the missing values.

You could use `isnull` to fill these values directly, but this can be tedious, especially if there are many values to replace. Since this is a common task in data science, pandas offers the `fillna` method, which creates a copy of the `Series` or `DataFrame` with the missing values replaced by a value of your choice. Let's create another example `Series` to see how this works in practice.


### Categorical Data (Non-numeric)
First, let's look at non-numeric data. In datasets, there are columns containing categorical data, such as Gender, True or False, etc.

In most cases, missing values in these columns are replaced with the `mode` of the column. For example, if we have 100 data points where 90 indicate True, 8 indicate False, and 2 are missing, we can fill the 2 missing values with True, based on the overall column data.

Once again, domain knowledge can be applied here. Let's consider an example of filling missing values using the mode.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

Now, let's first find the mode before filling the `None` value with the mode.


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

So, we will replace None with True


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

As we can see, the null value has been replaced. Needless to say, we could have written anything in place of `'True'` and it would have been substituted.


### Numeric Data
Now, let's talk about numeric data. There are two common methods for handling missing values in this case:

1. Replace with the median of the column
2. Replace with the mean of the column

We use the median when the data is skewed or contains outliers. This is because the median is less affected by outliers.

When the data is normalized, the mean can be used, as in such cases, the mean and median are usually quite similar.

To start, let's take a column that follows a normal distribution and fill its missing values with the mean of the column.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

The mean of the column is


In [33]:
np.mean(fill_with_mean[0])

0.0

Filling with mean


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

As we can see, the missing value has been replaced with its mean.


Now let us try another dataframe, and this time we will replace the None values with the median of the column.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

The median of the second column is


In [36]:
fill_with_median[1].median()

4.0

Filling with median


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

As we can see, the NaN value has been replaced by the median of the column


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

You can fill all of the null entries with a single value, such as `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> Key takeaways:
1. Missing values should be filled in when there is limited data or a clear strategy for handling the missing data.
2. Domain knowledge can help approximate and fill in missing values.
3. For categorical data, missing values are often replaced with the mode of the column.
4. For numeric data, missing values are typically filled with the mean (for normalized datasets) or the median of the column.


### Exercise:


In [40]:
# What happens if you try to fill null values with a string, like ''?


You can **forward-fill** null values, which is to use the last valid value to fill a null:


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

You can also **back-fill** to propagate the next valid value backward to fill a null:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

As you might guess, this works the same with DataFrames, but you can also specify an `axis` along which to fill null values:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

Notice that when a previous value is not available for forward-filling, the null value remains.


### Exercise:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


You can be creative about how you use `fillna`. For example, let's look at `example4` again, but this time let's fill the missing values with the average of all of the values in the `DataFrame`:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

Notice that column 3 still has no values: the default behavior is to fill values row by row.

> **Takeaway:** There are various approaches to handle missing values in your datasets. The method you choose (whether to remove them, replace them, or decide how to replace them) should depend on the specific characteristics of the data. As you work with and explore more datasets, you'll gain a better understanding of how to address missing values effectively.


### Encoding Categorical Data

Machine learning models can only process numerical data. They cannot differentiate between "Yes" and "No," but they can distinguish between 0 and 1. Therefore, after handling missing values, we need to convert categorical data into a numerical format that the model can interpret.

There are two main methods for encoding, which we will discuss next.


**LABEL ENCODING**

Label encoding essentially involves converting each category into a number. For instance, suppose we have a dataset of airline passengers, and there is a column indicating their class, which could be one of the following: ['business class', 'economy class', 'first class']. If we apply label encoding to this, it would be transformed into [0, 1, 2]. Let's look at an example using code. Since we will be learning `scikit-learn` in the upcoming notebooks, we won't use it here.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

To perform label encoding on the 1st column, we have to first describe a mapping from each class to a number, before replacing


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

As we can see, the output aligns with our expectations. So, when should we use label encoding? Label encoding is applied in one or both of the following scenarios:
1. When there are a large number of categories
2. When the categories have a specific order.


**ONE HOT ENCODING**

One Hot Encoding is another method of encoding. In this approach, each category in the column is represented as a separate column, and each data point is assigned a 0 or 1 depending on whether it belongs to that category. Thus, if there are n unique categories, n columns will be added to the dataframe.

For instance, consider the same airplane class example. The categories were: ['business class', 'economy class', 'first class']. If we apply one hot encoding, the dataset will have the following three additional columns: ['class_business class', 'class_economy class', 'class_first class'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Let us perform one hot encoding on the 1st column


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

Each one-hot encoded column contains 0 or 1, which specifies whether that category exists for that data point.


When do we use one hot encoding? One hot encoding is used in one or both of the following cases:

1. When the number of categories and the size of the dataset are small.
2. When the categories do not follow any specific order.


> Key Takeaways:
1. Encoding is used to transform non-numeric data into numeric data.
2. There are two types of encoding: Label encoding and One Hot encoding, which can be applied depending on the requirements of the dataset.


## Removing duplicate data

> **Learning goal:** By the end of this subsection, you should be comfortable identifying and removing duplicate values from DataFrames.

In addition to missing data, you will often encounter duplicated data in real-world datasets. Fortunately, pandas provides an easy way to detect and remove duplicate entries.


### Identifying duplicates: `duplicated`

You can easily identify duplicate values using the `duplicated` method in pandas, which provides a Boolean mask indicating whether an entry in a `DataFrame` is a duplicate of a previous one. Let's create another example `DataFrame` to demonstrate this.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### Dropping duplicates: `drop_duplicates`
`drop_duplicates` simply returns a copy of the data where all `duplicated` values are `False`:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

Both `duplicated` and `drop_duplicates` default to consider all columns, but you can specify that they examine only a subset of columns in your `DataFrame`:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **Takeaway:** Removing duplicate data is an essential part of almost every data-science project. Duplicate data can change the results of your analyses and give you inaccurate results!


## Real-World Data Quality Checks

> **Learning goal:** By the end of this section, you should feel confident identifying and addressing common real-world data quality issues, such as inconsistent categorical values, unusual numeric values (outliers), and duplicate entities with slight variations.

While missing values and exact duplicates are frequent problems, real-world datasets often present more nuanced challenges:

1. **Inconsistent categorical values**: The same category written differently (e.g., "USA", "U.S.A", "United States").
2. **Unusual numeric values**: Extreme outliers that may indicate data entry mistakes (e.g., age = 999).
3. **Near-duplicate rows**: Entries that represent the same entity but with minor differences.

Let’s dive into methods for detecting and resolving these issues.


### Creating a Sample "Dirty" Dataset

First, let's create a sample dataset that includes the types of problems we often face in real-world data:


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. Detecting Inconsistent Categorical Values

Notice that the `country` column contains multiple variations for the same countries. Let's pinpoint these inconsistencies:


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### Standardizing Categorical Values

We can create a mapping to standardize these values. A straightforward method is to convert them to lowercase and use a mapping dictionary:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**Alternative: Using Fuzzy Matching**

For more complex scenarios, the `rapidfuzz` library can be used to perform fuzzy string matching and automatically identify similar strings:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. Detecting Abnormal Numeric Values (Outliers)

Looking at the `age` column, we have some suspicious values such as 199 and -5. Let's apply statistical methods to identify these outliers.


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### Using IQR (Interquartile Range) Method

The IQR method is a reliable statistical approach for identifying outliers, as it is less affected by extreme values:


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### Using Z-Score Method

The Z-score method identifies outliers based on standard deviations from the mean:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### Handling Outliers

Once detected, outliers can be handled in several ways:
1. **Remove**: Delete rows containing outliers (if they are errors)
2. **Cap**: Replace them with boundary values
3. **Replace with NaN**: Treat them as missing data and apply imputation techniques
4. **Keep**: If they are valid extreme values


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. Identifying Near-Duplicate Rows

Observe that our dataset contains multiple entries for "John Smith" with slightly varying details. Let's pinpoint possible duplicates by analyzing name similarity.


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### Finding Near-Duplicates with Fuzzy Matching

For more advanced duplicate detection, fuzzy matching can be used to identify similar names:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### Handling Duplicates

Once identified, you need to decide how to handle duplicates:
1. **Keep the first occurrence**: Use `drop_duplicates(keep='first')`
2. **Keep the last occurrence**: Use `drop_duplicates(keep='last')`
3. **Aggregate information**: Combine information from duplicate rows
4. **Manual review**: Flag for human review


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### Summary: Complete Data Cleaning Pipeline

Let's combine everything into a full-fledged data cleaning pipeline:


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 Challenge Exercise

Now it's your turn! Below is a new row of data with multiple quality issues. Can you:

1. Identify all the issues in this row
2. Write code to clean each issue
3. Add the cleaned row to the dataset

Here's the problematic data:


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### Key Takeaways

1. **Inconsistent categories** are common in real-world data. Always review unique values and standardize them using mappings or fuzzy matching.

2. **Outliers** can have a significant impact on your analysis. Combine domain knowledge with statistical methods (IQR, Z-score) to identify them.

3. **Near-duplicates** are more challenging to detect than exact duplicates. Use techniques like fuzzy matching and data normalization (lowercasing, removing extra spaces) to spot them.

4. **Data cleaning is an iterative process**. You may need to apply several techniques and evaluate the results before finalizing your cleaned dataset.

5. **Document your decisions**. Record the cleaning steps you applied and the reasoning behind them, as this is crucial for reproducibility and transparency.

> **Best Practice:** Always retain a copy of your original "dirty" data. Avoid overwriting your source data files—create cleaned versions with clear naming conventions, such as `data_cleaned.csv`.



---

**Disclaimer**:  
This document has been translated using the AI translation service [Co-op Translator](https://github.com/Azure/co-op-translator). While we strive for accuracy, please note that automated translations may contain errors or inaccuracies. The original document in its native language should be regarded as the authoritative source. For critical information, professional human translation is recommended. We are not responsible for any misunderstandings or misinterpretations resulting from the use of this translation.
